README.md
# Net::SSH::Gateway

<em><b>Please note: this project is in maintenance mode. It is not under active development but pull requests are very much welcome. Just be sure to include tests!</b></em>

* Docs: http://net-ssh.github.com/net-ssh-gateway
* Issues: https://github.com/net-ssh/net-ssh-gateway/issues
* Codes: https://github.com/net-ssh/net-ssh-gateway
* Email: net-ssh@solutious.com

<em>As of v1.1.1, all gem releases are signed. See INSTALL.</em>


## DESCRIPTION:

Net::SSH::Gateway is a library for programmatically tunnelling connections to servers via a single "gateway" host. It is useful for establishing Net::SSH connections to servers behind firewalls, but can also be used to forward ports and establish connections of other types, like HTTP, to servers with restricted access.

## FEATURES:

* Easily manage forwarded ports
* Establish Net::SSH connections through firewalls

## SYNOPSIS:

In a nutshell:

  require 'net/ssh/gateway'

  gateway = Net::SSH::Gateway.new('host', 'user')

  gateway.ssh("host.private", "user") do |ssh|
    puts ssh.exec!("hostname")
  end

  gateway.open("host.private", 80) do |port|
    Net::HTTP.get_print("127.0.0.1", "/path", port)
  end

  gateway.shutdown!

  # As of 1.1.0, you can also specify the wait time for the
  # gateway thread with the :loop_wait option.
  gateway = Net::SSH::Gateway.new('host', 'user', :loop_wait => 0.001)

See Net::SSH::Gateway for more documentation.

## REQUIREMENTS:

* net-ssh (version 2)

If you want to run the tests or use any of the Rake tasks, you'll need:

* Echoe (for the Rakefile)
* Mocha (for the tests)

## INSTALL:

    $ gem install net-ssh-gateway

However, in order to be sure the code you're installing hasn't been tampered with, it's recommended that you verify the [signature](http://guides.rubygems.org/security/). To do this, you need to add the project's public key as a trusted certificate (you only need to do this once):

    # Add the public key as a trusted certificate
    # (You only need to do this once)
    $ curl -O https://raw.githubusercontent.com/net-ssh/net-ssh-gateway/master/net-ssh-public_cert.pem
    $ gem cert --add net-ssh-public_cert.pem

Then, when install the gem, do so with high security:

    $ gem install net-ssh-gateway -P HighSecurity

If you don't add the public key, you'll see an error like "Couldn't verify data signature".


## LICENSE:

(The MIT License)

Copyright (c) 2008 Jamis Buck <jamis@37signals.com>

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGES.txt

=== 1.3.0 / 26 Jan 2017

* Fix for loop_wait option on initialization [mfazekas, tpitale]
* Use bundler and remove remove jeweler [tpitale]
* Use minitest instead of test/unit [tpitale]
* Added Travis CI config [tpitale]


=== 1.2.0 / 06 Feb 2013

* Added public cert. All gem releases are now signed. See INSTALL in readme.

=== 1.1.0 / 30 Apr 2011

* Fix for sporadicly stalled SSH connections in Ruby 1.9 [Mat Trudel]
* Decrease wait time within Net::SSH session loop [Grant Hutchins, Michael Schubert]

=== 1.0.1 / 1 Feb 2009

* Allow local port to be specified [Will Klancnik]

=== 1.0.0 / 1 May 2008

* (no changes since the preview release)

=== 1.0 Preview Release 1 (0.99.0) / 10 Apr 2008

* First release of Net::SSH::Gateway


test/net/ssh/gateway_test.rb
# ruby -I../net-ssh/lib -Ilib -Itest -rrubygems test/gateway_test.rb
require 'bundler/setup'
require 'minitest/autorun'
require 'mocha/mini_test'
require 'net/ssh/gateway'

class GatewayTest < MiniTest::Test
  def teardown
    Thread.list { |t| t.kill unless Thread.current == t }
  end

  def test_shutdown_without_any_open_connections_should_terminate_session
    session, gateway = new_gateway
    session.expects(:close)
    gateway.shutdown!
    assert !gateway.active?
    assert session.forward.active_locals.empty?
  end

  def test_open_should_start_local_ports_at_65535
    gateway_session, gateway = new_gateway
    assert_equal 65535, gateway.open("app1", 22)
    assert_equal [65535, "app1", 22], gateway_session.forward.active_locals[65535]
  end

  def test_open_should_decrement_port_and_retry_if_ports_are_in_use
    gateway_session, gateway = new_gateway(:reserved => lambda { |n| n > 65000 })
    assert_equal 65000, gateway.open("app1", 22)
    assert_equal [65000, "app1", 22], gateway_session.forward.active_locals[65000]
  end

  def test_open_with_explicit_local_port_should_use_that_port
    gateway_session, gateway = new_gateway
    assert_equal 8181, gateway.open("app1", 22, 8181)
    assert_equal [8181, "app1", 22], gateway_session.forward.active_locals[8181]
  end

  def test_ssh_should_return_connection_when_no_block_is_given
    gateway_session, gateway = new_gateway
    expect_connect_to("127.0.0.1", "user", :port => 65535).returns(result = mock("session"))
    newsess = gateway.ssh("app1", "user")
    assert_equal result, newsess
    assert_equal [65535, "app1", 22], gateway_session.forward.active_locals[65535]
  end

  def test_ssh_with_block_should_yield_session_and_then_close_port
    gateway_session, gateway = new_gateway
    expect_connect_to("127.0.0.1", "user", :port => 65535).yields(result = mock("session"))
    yielded = false
    gateway.ssh("app1", "user") do |newsess|
      yielded = true
      assert_equal result, newsess
    end
    assert yielded
    assert gateway_session.forward.active_locals.empty?
  end

  def test_shutdown_should_cancel_active_forwarded_ports
    gateway_session, gateway = new_gateway
    gateway.open("app1", 80)
    assert !gateway_session.forward.active_locals.empty?
    gateway.shutdown!
    assert gateway_session.forward.active_locals.empty?
  end

  private

    def expect_connect_to(host, user, options={})
      Net::SSH.expects(:start).with do |real_host, real_user, real_options|
        host == real_host &&
        user == real_user &&
        options[:port] == real_options[:port]
      end
    end

    def new_gateway(options={})
      session = MockSession.new(options)
      expect_connect_to("test.host", "tester").returns(session)
      [session, Net::SSH::Gateway.new("test.host", "tester")]
    end

    class MockForward
      attr_reader :active_locals

      def initialize(options)
        @options = options
        @active_locals = {}
      end

      def cancel_local(port)
        @active_locals.delete(port)
      end

      def local(lport, host, rport)
        raise Errno::EADDRINUSE if @options[:reserved] && @options[:reserved][lport]
        @active_locals[lport] = [lport, host, rport]
      end
    end

    class MockSession
      attr_reader :forward

      def initialize(options={})
        @forward = MockForward.new(options)
      end

      def close
      end

      def process(wait=nil)
        true
      end

      def looping?
        @looping
      end

      def loop
        @looping = true
        sleep 0.1 while yield
        @looping = false
      end
    end
end


lib/net/ssh/gateway.rb
require 'thread'
require 'net/ssh'
require 'net/ssh/version'

# A Gateway is an object that allows you to tunnel network connections through
# a publicly visible host to a host hidden behind it. This is particularly
# useful when dealing with hosts behind a firewall. One host will generally
# be visible (and accessible) outside the firewall, while the others will all
# be behind the firewall, and the only way to access those restricted hosts
# is by first logging into the publicly visible host, and from thence logging
# into the restricted ones.
#
# This class makes it easy to programmatically connect to these restricted
# hosts via SSH. You can either simply forward a port from the local host to
# the remote host, or you can open a new Net::SSH connection to the remote
# host via a forwarded port.
#
#   require 'net/ssh/gateway'
#
#   gateway = Net::SSH::Gateway.new('host.name', 'user')
#
#   gateway.open('hidden.host', 80) do |port|
#     Net::HTTP.get_print '127.0.0.1', '/path', port
#   end
#
#   gateway.ssh('hidden.host', 'user') do |ssh|
#     puts ssh.exec!("hostname")
#   end
#
#   gateway.shutdown!
#
# Port numbers are allocated automatically, beginning at MAX_PORT and
# decrementing on each request for a new port until MIN_PORT is reached. If
# a port is already in use, this is detected and a different port will be
# assigned.
class Net::SSH::Gateway
  # A trivial class for representing the version of this library.
  class Version < Net::SSH::Version
    # The major component of the library's version
    MAJOR = 1

    # The minor component of the library's version
    MINOR = 2

    # The tiny component of the library's version
    TINY  = 0

    # The library's version as a Version instance
    CURRENT = new(MAJOR, MINOR, TINY)

    # The library's version as a String instance
    STRING = CURRENT.to_s
  end

  # The maximum port number that the gateway will attempt to use to forward
  # connections from.
  MAX_PORT = 65535

  # The minimum port number that the gateway will attempt to use to forward
  # connections from.
  MIN_PORT = 1024

  # Instantiate a new Gateway object, using the given remote host as the
  # tunnel. The arguments here are identical to those for Net::SSH.start, and
  # are passed as given to that method to start up the gateway connection.
  #
  #   gateway = Net::SSH::Gateway.new('host', 'user', :password => "password")
  #
  # As of 1.1 there is an additional option to specify the wait time for
  # the gateway thread. The default is 0.001 seconds and can be changed
  # with the :loop_wait option.
  #
  def initialize(host, user, options={})
    @loop_wait = options.delete(:loop_wait) || 0.001
    @session = Net::SSH.start(host, user, options)
    @session_mutex = Mutex.new
    @port_mutex = Mutex.new
    @next_port = MAX_PORT
    initiate_event_loop!
  end

  # Returns +true+ if the gateway is currently open and accepting connections.
  # This will be the case unless #shutdown! has been invoked.
  def active?
    @active
  end

  # Shuts down the gateway by closing all forwarded ports and then closing
  # the gateway's SSH session.
  def shutdown!
    return unless active?

    @session_mutex.synchronize do
      # cancel all active forward channels
      @session.forward.active_locals.each do |lport, host, port|
        @session.forward.cancel_local(lport)
      end
    end

    @active = false

    @thread.join
    @session.close
  end

  # Opens a new port on the local host and forwards it to the given host/port
  # via the gateway host. If a block is given, the newly allocated port
  # number will be yielded to the block, and the port automatically closed
  # (see #close) when the block finishes. Otherwise, the port number will be
  # returned, and the caller is responsible for closing the port (#close).
  #
  #   gateway.open('host', 80) do |port|
  #     # ...
  #   end
  #
  #   port = gateway.open('host', 80)
  #   # ...
  #   gateway.close(port)
  #
  # If +local_port+ is not specified, the next available port will be used.
  def open(host, port, local_port=nil)
    ensure_open!

    actual_local_port = local_port || next_port

    @session_mutex.synchronize do
      @session.forward.local(actual_local_port, host, port)
    end

    if block_given?
      begin
        yield actual_local_port
      ensure
        close(actual_local_port)
      end
    else
      return actual_local_port
    end
  rescue Errno::EADDRINUSE
    raise if local_port # if a local port was explicitly requested, bubble the error up
    retry
  end

  # Cancels port-forwarding over an open port that was previously opened via
  # #open.
  def close(port)
    ensure_open!

    @session_mutex.synchronize do
      @session.forward.cancel_local(port)
    end
  end

  # Forwards a new connection to the given +host+ and opens a new Net::SSH
  # connection to that host over the forwarded port. If a block is given,
  # the new SSH connection will be yielded to the block, and autoclosed
  # when the block terminates. The forwarded port will be autoclosed as well.
  # If no block was given, the new SSH connection will be returned, and it
  # is up to the caller to terminate both the connection and the forwarded
  # port when done.
  #
  #   gateway.ssh('host', 'user') do |ssh|
  #     # ...
  #   end
  #
  #   ssh = gateway.ssh('host', 'user')
  #   # ...
  #   ssh.close
  #   gateway.close(ssh.transport.port)
  def ssh(host, user, options={}, &block)
    local_port = open(host, options[:port] || 22)

    begin
      Net::SSH.start("127.0.0.1", user, options.merge(:port => local_port), &block)
    ensure
      close(local_port) if block || $!
    end
  end

  private

    # Raises a RuntimeError if the gateway is not active. This is used as a
    # sanity check to make sure a client doesn't try to call any methods on
    # a closed gateway.
    def ensure_open!
      raise "attempt to use a closed gateway" unless active?
    end

    # Fires up the gateway session's event loop within a thread, so that it
    # can run in the background. The loop will run for as long as the gateway
    # remains active.
    def initiate_event_loop!
      @active = true

      @thread = Thread.new do
        while @active
          @session_mutex.synchronize do
            @session.process(@loop_wait)
          end
          Thread.pass
        end
      end
    end

    # Grabs the next available port number and returns it.
    def next_port
      @port_mutex.synchronize do
        port = @next_port
        @next_port -= 1
        @next_port = MAX_PORT if @next_port < MIN_PORT
        port
      end
    end
end
